src/lib/field.ts

```diff
@@ -0,0 +1,794 @@
+import { Snarky, SnarkyField, Provable } from '../snarky.js';
+import { Field as Fp } from '../provable/field-bigint.js';
+import { Bool } from '../snarky.js';
+import { defineBinable } from '../bindings/lib/binable.js';
+import type { NonNegativeInteger } from '../bindings/crypto/non-negative.js';
+import { asProver } from './provable-context.js';
+
+export { Field, ConstantField, FieldType, FieldVar, FieldConst, isField };
+
+const SnarkyFieldConstructor = SnarkyField(1).constructor;
+
+type FieldConst = Uint8Array;
+
+function constToBigint(x: FieldConst): Fp {
+  return Fp.fromBytes([...x]);
+}
+function constFromBigint(x: Fp) {
+  return Uint8Array.from(Fp.toBytes(x));
+}
+
+const FieldConst = {
+  fromBigint: constFromBigint,
+  toBigint: constToBigint,
+  [0]: constFromBigint(0n),
+  [1]: constFromBigint(1n),
+  [-1]: constFromBigint(Fp(-1n)),
+};
+
+enum FieldType {
+  Constant,
+  Var,
+  Add,
+  Scale,
+}
+
+type FieldVar =
+  | [FieldType.Constant, FieldConst]
+  | [FieldType.Var, number]
+  | [FieldType.Add, FieldVar, FieldVar]
+  | [FieldType.Scale, FieldConst, FieldVar];
+
+type ConstantFieldVar = [FieldType.Constant, FieldConst];
+
+const FieldVar = {
+  constant(x: bigint | FieldConst): ConstantFieldVar {
+    let x0 = typeof x === 'bigint' ? FieldConst.fromBigint(x) : x;
+    return [FieldType.Constant, x0];
+  },
+  // TODO: handle (special) constants
+  add(x: FieldVar, y: FieldVar): FieldVar {
+    return [FieldType.Add, x, y];
+  },
+  // TODO: handle (special) constants
+  scale(x: FieldVar, c: FieldConst): FieldVar {
+    return [FieldType.Scale, c, x];
+  },
+  [0]: [0, FieldConst[0]] satisfies ConstantFieldVar,
+  [1]: [0, FieldConst[1]] satisfies ConstantFieldVar,
+  [-1]: [0, FieldConst[-1]] satisfies ConstantFieldVar,
+};
+
+type ConstantField = Field & { value: ConstantFieldVar };
+
+/**
+ * An element of a finite field.
+ */
+class Field {
+  value: FieldVar;
+
+  /**
+   * The field order as a `bigint`.
+   */
+  static ORDER = Fp.modulus;
+
+  /**
+   * Coerces anything field-like to a {@link Field}.
+   */
+  constructor(x: bigint | number | string | Field | FieldVar) {
+    if (Field.#isField(x)) {
+      this.value = x.value;
+      return;
+    }
+    // fieldVar
+    if (Array.isArray(x)) {
+      this.value = x;
+      return;
+    }
+    // TODO this should handle common values efficiently by reading from a lookup table
+    this.value = FieldVar.constant(Fp(x));
+  }
+
+  // helpers
+  static #isField(x: bigint | number | string | Field | FieldVar): x is Field {
```

Comment:
I'm using the new JS private class fields with `#` to encapsulate logic. Right now, these are compiled to a polyfill by TS, but once we switch to es2022 target the builtin JS feature is used